src/h1/mod.rs

```diff
@@ -134,72 +134,84 @@ impl HttpClient for H1Client {
+
+            match scheme {
+                "http" => {
+                    let pool_ref = if let Some(pool_ref) = self.http_pools.get(&addr) {
+                        pool_ref
+                    } else {
+                        let manager = TcpConnection::new(addr);
+                        let pool = Pool::<TcpStream, std::io::Error>::new(
+                            manager,
+                            self.max_concurrent_connections,
+                        );
+                        self.http_pools.insert(addr, pool);
+                        self.http_pools.get(&addr).unwrap()
+                    };
+
+                    // Deadlocks are prevented by cloning an inner pool Arc and dropping the original locking reference before we await.
+                    let pool = pool_ref.clone();
+                    std::mem::drop(pool_ref);
+
+                    let stream = match pool.get().await {
+                        Ok(s) => s,
+                        Err(_) if has_another_addr => continue,
+                        Err(e) => return Err(Error::from_str(400, e.to_string()))?,
+                    };
+
+                    req.set_peer_addr(stream.peer_addr().ok());
+                    req.set_local_addr(stream.local_addr().ok());
+                    return client::connect(TcpConnWrapper::new(stream), req).await;
+                }
+                #[cfg(any(feature = "native-tls", feature = "rustls"))]
+                "https" => {
+                    let pool_ref = if let Some(pool_ref) = self.https_pools.get(&addr) {
+                        pool_ref
+                    } else {
+                        let manager = TlsConnection::new(host.clone(), addr);
+                        let pool = Pool::<TlsStream<TcpStream>, Error>::new(
+                            manager,
+                            self.max_concurrent_connections,
+                        );
+                        self.https_pools.insert(addr, pool);
+                        self.https_pools.get(&addr).unwrap()
+                    };
+
+                    // Deadlocks are prevented by cloning an inner pool Arc and dropping the original locking reference before we await.
+                    let pool = pool_ref.clone();
+                    std::mem::drop(pool_ref);
+
+                    let stream = match pool.get().await {
+                        Ok(s) => s,
+                        Err(_) if has_another_addr => continue,
+                        Err(e) => return Err(Error::from_str(400, e.to_string()))?,
+                    };
+
+                    req.set_peer_addr(stream.get_ref().peer_addr().ok());
+                    req.set_local_addr(stream.get_ref().local_addr().ok());
+
+                    return client::connect(TlsConnWrapper::new(stream), req).await;
+                }
+                _ => unreachable!(),
             }
-            _ => unreachable!(),
         }
+
+        Err(Error::from_str(
```

Comment:
(Isn't this also unreachable?)

Reply:
Not really; we may not have addressed and then the `for` would be skipped.